transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java

```diff
@@ -196,75 +214,82 @@ public ChannelFuture bind() {
     /**
      * Create a new {@link Channel} and bind it.
      */
-    public ChannelFuture bind(int inetPort) {
+    public Future<Channel> bind(int inetPort) {
         return bind(new InetSocketAddress(inetPort));
     }
 
     /**
      * Create a new {@link Channel} and bind it.
      */
-    public ChannelFuture bind(String inetHost, int inetPort) {
+    public Future<Channel> bind(String inetHost, int inetPort) {
         return bind(SocketUtils.socketAddress(inetHost, inetPort));
     }
 
     /**
      * Create a new {@link Channel} and bind it.
      */
-    public ChannelFuture bind(InetAddress inetHost, int inetPort) {
+    public Future<Channel> bind(InetAddress inetHost, int inetPort) {
         return bind(new InetSocketAddress(inetHost, inetPort));
     }
 
     /**
      * Create a new {@link Channel} and bind it.
      */
-    public ChannelFuture bind(SocketAddress localAddress) {
+    public Future<Channel> bind(SocketAddress localAddress) {
         validate();
         requireNonNull(localAddress, "localAddress");
         return doBind(localAddress);
     }
 
-    private ChannelFuture doBind(final SocketAddress localAddress) {
-        final ChannelFuture regFuture = initAndRegister();
-        final Channel channel = regFuture.channel();
+    private Future<Channel> doBind(final SocketAddress localAddress) {
+        EventLoop loop = group.next();
+        final Future<Channel> regFuture = initAndRegister(loop);
         if (regFuture.cause() != null) {
             return regFuture;
         }
 
+        Promise<Channel> bindPromise = new DefaultPromise<>(loop);
         if (regFuture.isDone()) {
             // At this point we know that the registration was complete and successful.
+            Channel channel = regFuture.getNow();
             ChannelPromise promise = channel.newPromise();
+            cascadeChannel(promise, bindPromise);
             doBind0(regFuture, channel, localAddress, promise);
-            return promise;
         } else {
             // Registration future is almost always fulfilled already, but just in case it's not.
-            final ChannelPromise promise = channel.newPromise();
-            regFuture.addListener((ChannelFutureListener) future -> {
+            regFuture.addListener((GenericFutureListener<Future<Channel>>) future -> {
```

Comment:
FYI: It looks like generics on GenericFutureListener are annoying, as you need to repeat Future. But worse, they are wrong, as there is no verification that the generic type matches the actual future's type, so it is an implicit cast (the implementation [hides the cast](https://github.com/netty/netty/blob/b6da6f58282b3e75417569e5a37dda85edbc7ca5/common/src/main/java/io/netty/util/concurrent/DefaultPromise.java#L576) by dropping generics).

Something else to clean up (separately) if we are going to be relying on these APIs more heavily?

Reply:
Yes I think we need to cleanup the generic stuff as a followup

Reply:
Seems this fits well under https://github.com/netty/netty/issues/8515's scope. Added a comment there.